starky/src/fibonacci_stark.rs

```diff
@@ -0,0 +1,98 @@
+use std::marker::PhantomData;
+
+use plonky2::field::extension_field::{Extendable, FieldExtension};
+use plonky2::field::packed_field::PackedField;
+use plonky2::hash::hash_types::RichField;
+use plonky2::plonk::circuit_builder::CircuitBuilder;
+
+use crate::constraint_consumer::{ConstraintConsumer, RecursiveConstraintConsumer};
+use crate::stark::Stark;
+use crate::vars::{StarkEvaluationTargets, StarkEvaluationVars};
+
+/// Toy STARK system used for testing.
+/// Computes a Fibonacci sequence with inital values `x0, x1` using the transition
+/// `x0 <- x1, x1 <- x0 + x1`.
+pub struct FibonacciStark<F: RichField + Extendable<D>, const D: usize> {
+    x0: F,
+    x1: F,
+    _phantom: PhantomData<F>,
+}
+
+impl<F: RichField + Extendable<D>, const D: usize> FibonacciStark<F, D> {
+    const NUM_COLUMNS: usize = 2;
+    const NUM_ROWS: usize = 1 << 5;
+
+    fn new(x0: F, x1: F) -> Self {
+        Self {
+            x0,
+            x1,
+            _phantom: PhantomData,
+        }
+    }
+
+    fn generate_trace(&self) -> Vec<[F; Self::NUM_COLUMNS]> {
+        (0..Self::NUM_ROWS)
+            .scan([self.x0, self.x1], |acc, _| {
+                let tmp = *acc;
+                acc[0] = tmp[1];
+                acc[1] = tmp[0] + tmp[1];
+                Some(tmp)
+            })
+            .collect()
+    }
+}
+
+impl<F: RichField + Extendable<D>, const D: usize> Stark<F, D> for FibonacciStark<F, D> {
+    const COLUMNS: usize = Self::NUM_COLUMNS;
+    const PUBLIC_INPUTS: usize = 0;
+
+    fn eval_packed_generic<FE, P, const D2: usize>(
+        &self,
+        vars: StarkEvaluationVars<FE, P, { Self::COLUMNS }, { Self::PUBLIC_INPUTS }>,
+        yield_constr: &mut ConstraintConsumer<P>,
+    ) where
+        FE: FieldExtension<D2, BaseField = F>,
+        P: PackedField<Scalar = FE>,
+    {
```

Comment:
Let's enforce the base case also -
```rust
        yield_constr.one_first_row(vars.local_values[0] - P::from(FE::from_basefield(self.x0)));
        yield_constr.one_first_row(vars.local_values[1] - P::from(FE::from_basefield(self.x1)));
```

Reply:
Oh I was thinking of using public inputs for that (once they're taken care of).

Reply:
Ok that works. I think we could go ahead and add constraints for public inputs, as per below comment, even though there's no verification yet

Reply:
Not that it really matters since it's a test, but on second thought since `x0`, `x1` are fields of the `Stark`, to me that makes it seem like they'd be baked into the constraint system.

If the `Stark` is more general and can take any `x0`, `x1` are public inputs, maybe it'd make sense to make them arguments to `generate_trace`, instead of fields?

Reply:
Good point! I'll implement that.